tendermint/src/error.rs

```diff
@@ -1,206 +1,213 @@
 //! Error types
 
-use anomaly::{BoxError, Context};
-use thiserror::Error;
-
 use crate::account;
 use crate::vote;
+use flex_error::*;
 
-/// Error type
-pub type Error = BoxError;
+pub type Error = anyhow::Error;
 
-/// Kinds of errors
-#[derive(Clone, Eq, PartialEq, Debug, Error)]
-pub enum Kind {
-    /// Cryptographic operation failed
-    #[error("cryptographic error")]
-    Crypto,
 
-    /// Malformatted or otherwise invalid cryptographic key
-    #[error("invalid key")]
-    InvalidKey,
+define_error! { KindError;
+    Crypto
+    [DisplayError<Error>]
+    |_| { format_args!("cryptographic error") },
 
-    /// Input/output error
-    #[error("I/O error")]
-    Io,
+    InvalidKey
+    [DisplayError<Error>]
+    |_| { format_args!("invalid key") },
 
-    /// Length incorrect or too long
-    #[error("length error")]
-    Length,
+    Io
+    [DisplayError<Error>]
+    |_| { format_args!("I/O error") },
 
-    /// Parse error
-    #[error("parse error")]
-    Parse,
+    Length
+    [DisplayError<Error>]
+    |_| { format_args!("length error") },
 
-    /// Network protocol-related errors
-    #[error("protocol error")]
-    Protocol,
+    Parse
+    [DisplayError<Error>]
```

Comment:
We only need to wrap upstream errors if they are coming from code not belong to the us. If you just want to attach additional details in string, you can instead define suberrors such as `Parse { reason: String }` and omit the error source.

Better yet, we should split out the different suberrors such as:

```rust
ParseUrl
  { url: String }
  [ DisplayError<url::ParseError> ]
  | e | { format_args!("error parsing url {}", e.url) }
InvalidHashLength
  { length: usize }
  | e | { format_args!("hash has invalid length {}", e.length) }
```